Tests/SwiftGRPCNIOTests/ServerThrowingTests.swift

```diff
@@ -0,0 +1,152 @@
+  }
+}
+
+/// Ensures that fulfilling the status promise (where possible) with an error yields the same result as failing the future.
+private class ErrorReturningEchoProviderNIO: ImmediateThrowingEchoProviderNIO {
+  // There's no status promise to fulfill for unary calls (only the response promise), so that case is omitted.
+  
+  override func expand(request: Echo_EchoRequest, context: StreamingResponseCallContext<Echo_EchoResponse>) -> EventLoopFuture<GRPCStatus> {
+    return context.eventLoop.newSucceededFuture(result: expectedError)
+  }
+  
+  override func collect(context: UnaryResponseCallContext<Echo_EchoResponse>) -> EventLoopFuture<(StreamEvent<Echo_EchoRequest>) -> Void> {
+    return context.eventLoop.newSucceededFuture(result: { _ in
+      context.responseStatus = expectedError
+      context.responsePromise.succeed(result: Echo_EchoResponse())
+    })
+  }
+  
+  override func update(context: StreamingResponseCallContext<Echo_EchoResponse>) -> EventLoopFuture<(StreamEvent<Echo_EchoRequest>) -> Void> {
+    return context.eventLoop.newSucceededFuture(result: { _ in
+      context.statusPromise.succeed(result: expectedError)
+    })
+  }
+}
+
+class ServerThrowingTests: NIOBasicEchoTestCase {
+  override func makeEchoProvider() -> Echo_EchoProvider_NIO { return ImmediateThrowingEchoProviderNIO() }
+  
+  static var allTests: [(String, (ServerThrowingTests) -> () throws -> Void)] {
+    return [
+      ("testUnary", testUnary),
+      ("testClientStreaming", testClientStreaming),
+      ("testServerStreaming", testServerStreaming),
+      ("testBidirectionalStreaming", testBidirectionalStreaming),
+    ]
+  }
+}
+
+class ServerDelayedThrowingTests: ServerThrowingTests {
+  override func makeEchoProvider() -> Echo_EchoProvider_NIO { return DelayedThrowingEchoProviderNIO() }
+}
+
+class ClientThrowingWhenServerReturningErrorTests: ServerThrowingTests {
+  override func makeEchoProvider() -> Echo_EchoProvider_NIO { return ErrorReturningEchoProviderNIO() }
+}
+
+extension ServerThrowingTests {
+  func testUnary() throws {
+    let call = client.get(Echo_EchoRequest(text: "foo"))
+    XCTAssertEqual(expectedError, try call.status.wait())
+    XCTAssertThrowsError(try call.response.wait()) {
+      XCTAssertEqual(expectedError, $0 as? GRPCStatus)
+    }
+  }
+  
+  func testClientStreaming() {
+    let call = client.collect()
+    XCTAssertNoThrow(try call.sendEnd().wait())
+    XCTAssertEqual(expectedError, try call.status.wait())
+    
+    if type(of: makeEchoProvider()) != ErrorReturningEchoProviderNIO.self {
+      // With `ErrorReturningEchoProviderNIO` we actually _return_ a response, which means that the `response` future
+      // will _not_ fail.
+      XCTAssertThrowsError(try call.response.wait()) {
+        XCTAssertEqual(expectedError, $0 as? GRPCStatus)
+      }
+    }
+  }
+  
+  func testServerStreaming() {
+    let call = client.expand(Echo_EchoRequest(text: "foo")) { _ in XCTFail("no message expected") }
```

Comment:
Could be useful to output the message in the `XCTFail("no message expected, got \($0)")`